homeassistant/components/ezviz/__init__.py

```diff
@@ -50,54 +64,73 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
-            # Fetch Entry id of main account and reload it.
-            for item in hass.config_entries.async_entries():
-                if item.data.get(CONF_TYPE) == ATTR_TYPE_CLOUD:
-                    _LOGGER.info("Reload Ezviz integration with new camera rtsp entry")
-                    await hass.config_entries.async_reload(item.entry_id)
+    if PLATFORMS_BY_TYPE[sensor_type]:
+
+        # Get user account token if not present.
+        if not entry.data.get(CONF_SESSION_ID):
+
+            try:
+                ezviz_client = await _get_ezviz_client_instance(hass, entry)
 
-        return True
+            except (InvalidURL, HTTPError, PyEzvizError) as error:
+                _LOGGER.error("Unable to connect to Ezviz service: %s", str(error))
+                raise ConfigEntryNotReady from error
 
-    try:
-        ezviz_client = await hass.async_add_executor_job(
-            _get_ezviz_client_instance, entry
+        if not ezviz_client:
+            # No Ezviz login session, call api login().
+
+            ezviz_client = EzvizClient(
+                token={
+                    CONF_SESSION_ID: entry.data.get(CONF_SESSION_ID),
+                    CONF_RFSESSION_ID: entry.data.get(CONF_RFSESSION_ID),
+                    "api_url": entry.data.get(CONF_URL),
+                },
+                timeout=entry.options.get(CONF_TIMEOUT, DEFAULT_TIMEOUT),
+            )
+
+            try:
+                await hass.async_add_executor_job(ezviz_client.login)
+
+            except (EzvizAuthTokenExpired, EzvizAuthVerificationCode) as error:
+                raise ConfigEntryAuthFailed from error
+
+            except (InvalidURL, HTTPError, PyEzvizError) as error:
+                _LOGGER.error("Unable to connect to Ezviz service: %s", str(error))
+                raise ConfigEntryNotReady from error
+
+        coordinator = EzvizDataUpdateCoordinator(
+            hass, api=ezviz_client, api_timeout=entry.options[CONF_TIMEOUT]
         )
-    except (InvalidURL, HTTPError, PyEzvizError) as error:
-        _LOGGER.error("Unable to connect to Ezviz service: %s", str(error))
-        raise ConfigEntryNotReady from error
 
-    coordinator = EzvizDataUpdateCoordinator(
-        hass, api=ezviz_client, api_timeout=entry.options[CONF_TIMEOUT]
-    )
-    await coordinator.async_refresh()
+        hass.data[DOMAIN][entry.entry_id] = {DATA_COORDINATOR: coordinator}
 
-    if not coordinator.last_update_success:
-        raise ConfigEntryNotReady
+        await coordinator.async_config_entry_first_refresh()
 
-    undo_listener = entry.add_update_listener(_async_update_listener)
+    entry.async_on_unload(entry.add_update_listener(_async_update_listener))
 
-    hass.data[DOMAIN][entry.entry_id] = {
-        DATA_COORDINATOR: coordinator,
-        DATA_UNDO_UPDATE_LISTENER: undo_listener,
-    }
-    hass.config_entries.async_setup_platforms(entry, PLATFORMS)
+    if sensor_type == ATTR_TYPE_CAMERA:
+        if hass.data.get(DOMAIN):
```

Comment:
Merge these 2 if-statements into 1

Reply:
Also why guard for `hass.data.get(DOMAIN)` ? 

You also can use `hass.data[DOMAIN]` because you set it to a dictionary above

Reply:
Totally correct. Removed redundant check.